openfe/protocols/openmm_rfe/equil_rfe_methods.py

```diff
@@ -339,179 +379,90 @@ def run(self, *, dry=False, verbose=True,
-        stateA_ligand_nchains = stateA_ligand_topology.getNumChains()
-        stateA_ligand_chain_id = stateA_modeller.topology.getNumChains()
-
-        # 4. Solvate the complex in a `concentration` mM cubic water box with
-        # `solvent_padding` from the solute to the edges of the box
-        if 'solvent' in stateA.components:
-            conc = stateA['solvent'].ion_concentration
-            pos = stateA['solvent'].positive_ion
-            neg = stateA['solvent'].negative_ion
-
-            stateA_modeller.addSolvent(
-                omm_forcefield_stateA,
-                model=solvation_settings.solvent_model,
-                padding=to_openmm(solvation_settings.solvent_padding),
-                positiveIon=pos, negativeIon=neg,
-                ionicStrength=to_openmm(conc),
-            )
+            ffcache = None
+
+        system_generator = system_creation.get_system_generator(
+            forcefield_settings=forcefield_settings,
+            thermo_settings=thermo_settings,
+            system_settings=system_settings,
+            cache=ffcache,
+            has_solvent=solvent_comp is not None,
+        )
 
-        # 5.  Create OpenMM system + topology + initial positions for "A" system
-        #  a. Get nonbond method
-        nonbonded_method = {
-            'pme': app.PME,
-            'nocutoff': app.NoCutoff,
-            'cutoffnonperiodic': app.CutoffNonPeriodic,
-            'cutoffperiodic': app.CutoffPeriodic,
-            'ewald': app.Ewald
-        }[system_settings.nonbonded_method.lower()]
-
-        #  b. Get the constraint method
-        constraints = {
-            'hbonds': app.HBonds,
-            'none': None,
-            'allbonds': app.AllBonds,
-            'hangles': app.HAngles
-            # vvv can be None so string it
-        }[str(forcefield_settings.constraints).lower()]
-
-        #  c. create the stateA System
-        stateA_system = omm_forcefield_stateA.createSystem(
-            stateA_modeller.topology,
-            nonbondedMethod=nonbonded_method,
-            nonbondedCutoff=to_openmm(system_settings.nonbonded_cutoff),
-            constraints=constraints,
-            rigidWater=forcefield_settings.rigid_water,
-            hydrogenMass=forcefield_settings.hydrogen_mass * omm_unit.amu,
-            removeCMMotion=forcefield_settings.remove_com,
+        # b. force the creation of paramaters
+        # Note: by default this is cached to ctx.shared/db.json so shouldn't
+        # incur too large a cost
+        for comp, offmol in small_mols.items():
+            system_generator.create_system(offmol.to_topology().to_openmm(),
+                                           molecules=[offmol])
+            if comp == mapping.componentA:
+                molB = mapping.componentB.to_openff()
+                system_generator.create_system(molB.to_topology().to_openmm(),
+                                               molecules=[molB])
```

Comment:
what's the benefit of forcing early creation of parameters here?

Reply:
It's useful when you have several concurrently running units (which I was trialing out with the solvation free energies), but I guess not so much here - I can remove.

Reply:
Nope - I remember exactly why this is needed now. Without the templates generated _a priori_, `modeller.addSolvent` stumbles upon itself. I.e. You need to create the force fields first, and then solvate.